### PR TITLE
able to mapper same column names when append

### LINKED ISSUE #1425: 'Value cannot be null' exception on trivial estimator append
The following code throws:

```csharp
      var env = new ConsoleEnvironment(seed: 1, conc: 0);
      var dataView = env.CreateDataView(new[]{new {A = "A", B = "B"}}).AssertStatic(env, c => (
        A: c.Text.Scalar,
        B: c.Text.Scalar));

      dataView.MakeNewEstimator().Append(row => (row.A, row.B));
```

If this is intentional feel free to close it. Though I found the error at least hard to understand. I needed this step for weird reasons.

Adding some simple op will make it work:

```csharp
      dataView.MakeNewEstimator().Append(row => (row.A, row.B, c: row.B));
```

### PR DESCRIPTION
**Pull request by feiyun0112**
Fix #1425

### CODE REVIEW

**Comment by Zruty0** [comment]
Thanks @feiyun0112 for your contribution!

File: src/Microsoft.ML.Data/DataLoadSave/EstimatorChain.cs [hunk 1]
```diff
@@ -71,7 +71,12 @@ public SchemaShape GetOutputSchema(SchemaShape inputSchema)
         public EstimatorChain<TNewTrans> Append<TNewTrans>(IEstimator<TNewTrans> estimator, TransformerScope scope = TransformerScope.Everything)
             where TNewTrans : class, ITransformer
         {
-            Contracts.CheckValue(estimator, nameof(estimator));
+            if(estimator==null)
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/DataLoadSave/EstimatorChain.cs]: >if(estimator==null) [](start = 12, length = 19)

if you use Visual Studio, press `Ctrl-k-d` to fix the spacing. Otherwise, please fix the spacing :)

File: src/Microsoft.ML.Data/StaticPipe/Estimator.cs [hunk 2]
```diff
@@ -69,7 +69,6 @@ string NameMap(PipelineColumn col)
 
                 var readerEst = StaticPipeUtils.GeneralFunctionAnalyzer(Env, ch, input, fakeReconciler, mapper, out var estTail, NameMap);
                 ch.Assert(readerEst == null);
-                ch.AssertValue(estTail);
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/StaticPipe/Estimator.cs]: >ch.AssertValue(estTail); [](start = 16, length = 24)

I think the correct fix is NOT to allow `nulls` through all of our estimators' `Append` methods, but to have an `if` statement here to only append new estimator if `estTail` is not `null`

**Comment by feiyun0112** [comment]
@Zruty0 
I have changed codes, please review, thanks!

File: src/Microsoft.ML.Data/DataLoadSave/EstimatorChain.cs [hunk 3]
```diff
@@ -72,7 +72,14 @@ public EstimatorChain<TNewTrans> Append<TNewTrans>(IEstimator<TNewTrans> estimat
             where TNewTrans : class, ITransformer
         {
             Contracts.CheckValue(estimator, nameof(estimator));
+
             return new EstimatorChain<TNewTrans>(_estimators.AppendElement(estimator), _scopes.AppendElement(scope));
         }
+
+        public EstimatorChain<TNewTrans> Append<TNewTrans>()
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/DataLoadSave/EstimatorChain.cs]: >Append [](start = 41, length = 6)

Why is this needed? And why does it create a new class?

File: src/Microsoft.ML.Data/StaticPipe/Estimator.cs (hunk 2 above)
Comment (Zruty0) [reply src/Microsoft.ML.Data/StaticPipe/Estimator.cs]: how about `var est = estTail == null ? AsDynamic : AsDynamic:Append(estTail)` ?

---
In reply to: [229018643](https://github.com/dotnet/machinelearning/pull/1426#discussion_r229018643) [](ancestors = 229018643)

**Comment by Zruty0** [comment]
Did one more round. I am sorry, I was previously not clear enough of what I wanted :)

We don't really want to allow appending 'nothing' to an estimator chain. The calling code is responsible for checking whether append is warranted or not.

---
In reply to: [434275642](https://github.com/dotnet/machinelearning/pull/1426#issuecomment-434275642) [](ancestors = 434275642)

File: src/Microsoft.ML.Data/DataLoadSave/EstimatorChain.cs (hunk 3 above)
Comment (Zruty0) [reply src/Microsoft.ML.Data/DataLoadSave/EstimatorChain.cs]: Actually, this looks like a crude way to re-map the existing chain 'that ends with Foo' to another chain 'that ends with Bar'. I don't think this functionality is good, or useful in this case.

---
In reply to: [229384750](https://github.com/dotnet/machinelearning/pull/1426#discussion_r229384750) [](ancestors = 229384750)

**Comment by feiyun0112** [comment]
> We don't really want to allow appending 'nothing' to an estimator chain. The calling code is responsible for checking whether append is warranted or not.

OK, close this PR
